File: src/interception/screeningPipeline.ts
import { IntegratedScanner } from "../integrated-scanner/integratedScanner.js";
import type { ThreatReport } from "../analyzer/reportGenerator.js";
import type { Severity } from "../types.js";
import { decodeTransaction, transactionToScanTarget, type DecodedTransaction } from "./txDecoder.js";

export type ScreeningDecision = "allow" | "warn" | "block";

export interface ScreeningResult {
  decision: ScreeningDecision;
  severity: Severity;
  score: number;
  decoded: DecodedTransaction;
  report: ThreatReport;
}

export interface ScreeningPipelineOptions {
  scanner?: IntegratedScanner;
  /** Severity at/above which a transaction is blocked outright (default "critical"). */
  blockAt?: Severity;
  /** Severity at/above which the user must explicitly confirm (default "high"). */
  warnAt?: Severity;
}

const RANK: Record<Severity, number> = { low: 0, medium: 1, high: 2, critical: 3 };

/**
 * Transaction screening pipeline (PRD §5.5 / §10.1–10.2, FR-2).
 *
 * Decodes an intercepted signing request, runs it through the
 * {@link IntegratedScanner}, and maps the risk to an allow / warn / block
 * decision before any signature is produced.
 */
export class ScreeningPipeline {
  private readonly scanner: IntegratedScanner;
  private readonly blockAt: Severity;
  private readonly warnAt: Severity;

  constructor(options: ScreeningPipelineOptions = {}) {
    this.scanner = options.scanner ?? new IntegratedScanner();
    this.blockAt = options.blockAt ?? "critical";
    this.warnAt = options.warnAt ?? "high";
  }

  /** Screen a raw serialized transaction (signature vector + message, or message). */
  async screen(txBytes: Uint8Array, id?: string): Promise<ScreeningResult> {
    const decoded = decodeTransaction(txBytes);
    const target = transactionToScanTarget(decoded, id);
    const result = await this.scanner.scan(target);

    let decision: ScreeningDecision = "allow";
    if (RANK[result.severity] >= RANK[this.blockAt]) decision = "block";
    else if (RANK[result.severity] >= RANK[this.warnAt]) decision = "warn";

    return {
      decision,
      severity: result.severity,
      score: result.score,
      decoded,
      report: result.report,
    };
  }
}

File: android/engine/src/main/kotlin/com/daemonblockint/sync/engine/interception/ScreeningPipeline.kt
package com.daemonblockint.sync.engine.interception

import com.daemonblockint.sync.engine.Severity
import com.daemonblockint.sync.engine.analyzer.ThreatReport
import com.daemonblockint.sync.engine.integrated.IntegratedScanner
import com.daemonblockint.sync.engine.integrated.ScanResult

enum class ScreeningDecision { ALLOW, WARN, BLOCK }

data class ScreeningResult(
    val decision: ScreeningDecision,
    val severity: Severity,
    val score: Int,
    val decoded: DecodedTransaction,
    val report: ThreatReport,
)

data class ScreeningPipelineOptions(
    val scanner: IntegratedScanner? = null,
    val blockAt: Severity = Severity.CRITICAL,
    val warnAt: Severity = Severity.HIGH,
)

/**
 * Transaction screening pipeline (PRD §5.5 / §10.1–10.2, FR-2).
 *
 * Decodes an intercepted signing request, runs it through the
 * IntegratedScanner, and maps the risk to an allow / warn / block
 * decision before any signature is produced.
 */
class ScreeningPipeline(
    private val scanner: IntegratedScanner = IntegratedScanner(),
    private val blockAt: Severity = Severity.CRITICAL,
    private val warnAt: Severity = Severity.HIGH,
) {
    constructor(options: ScreeningPipelineOptions) : this(
        scanner = options.scanner ?: IntegratedScanner(),
        blockAt = options.blockAt,
        warnAt = options.warnAt,
    )

    /** Screen a raw serialized transaction (signature vector + message, or message). */
    fun screen(txBytes: ByteArray, id: String? = null): ScreeningResult {
        val decoded = decodeTransaction(txBytes)
        val target = transactionToScanTarget(decoded, id)
        val result = scanner.scanLocal(target)

        var decision = ScreeningDecision.ALLOW
        if ((Severity.rank[result.severity] ?: 0) >= (Severity.rank[blockAt] ?: 0))
            decision = ScreeningDecision.BLOCK
        else if ((Severity.rank[result.severity] ?: 0) >= (Severity.rank[warnAt] ?: 0))
            decision = ScreeningDecision.WARN

        return ScreeningResult(
            decision = decision,
            severity = result.severity,
            score = result.score,
            decoded = decoded,
            report = result.report,
        )
    }
}
